### ISSUE TITLE
stc fails to find Intrinsic string type in some cases

### ISSUE THREAD

**Issue by awareness481** [dudykr/stc#771]
Given the following code

```ts
type A = Uppercase<42>;
```

if you run `stc` on it you'll notice that stc does not report an error. Upon further investigation it looks like `stc` fails to find the built-in type `Uppercase` in this case. I think it is because `self.config.builtin` is set to `false`.

Note that it only fails to find the type in some cases, in most cases (like then `Uppercase` is a type of a function param, stc finds the type)

**Comment by awareness481** [comment]
As far as I know this should always be an error

```ts
type Uppercase = string;
//   ?^ Duplicate identifier 'Uppercase'.
```

**unless** https://www.typescriptlang.org/tsconfig#noLib is enabled. So perhaps something like that should work for all TS builtins? 

```
type Uppercase = string;
//   ?^ no error, noLib: true
```

If `noLib` is out of scope for `0.0.1` it could always be assumed that defining a type with has the same identifier as a TS builtin is an error

**Comment by kdy1** [comment]
I'm not sure about `noLib`, but I think we can exclude it for now.
I don't expect someone to use it for a real-world project

**Comment by awareness481** [comment]
Completely agree with `noLib` being excluded, but as far as the issue goes, I think I got confused because I was using `type Uppercase = string` to be debug, but the `type` keyword is not implemented yet right? Which is why it said `Uppercase missing type`.

I think I should close the issue.

**Comment by kdy1** [comment]
Type aliases are implemented

**Comment by kdy1** [comment]
But I'm confused too. This is about `Uppercase<42>`, right?
`type Uppercase = string` seems not relevant

**Comment by awareness481** [comment]
> But I'm confused too. This is about `Uppercase<42>`, right? `type Uppercase = string` seems not relevant

Yes, but in that case it fails to find `type A` and so it doesn't check `Uppercase<42>` AFAICT

**Comment by awareness481** [comment]
> ```ts
> type A = Uppercase<42>;
> ```

Maybe it's a bug with type aliases then, here's the log with the only `type A = Uppercase<42>;`

```rs
  DEBUG [(0)/types] Registering: A#2
    at crates/stc_ts_file_analyzer/src/analyzer/scope/mod.rs:791
    in register_type
    in Stmt with line_col: "(33:1-33:24)"

  DEBUG (0) Scope.find_type(`A#2`)
    at crates/stc_ts_file_analyzer/src/analyzer/scope/mod.rs:1192
    in find_local_type with name: A#2
    in find_type with name: A#2
    in register_type
    in Stmt with line_col: "(33:1-33:24)"

  DEBUG Analyzer.find_type('A#2')
    at crates/stc_ts_file_analyzer/src/analyzer/scope/mod.rs:1891
    in Scope::find_type with name: A#2
    in find_local_type with name: A#2
    in find_type with name: A#2
    in register_type
    in Stmt with line_col: "(33:1-33:24)"

  DEBUG Scope.find_type: failed to find type 'A#2'
    at crates/stc_ts_file_analyzer/src/analyzer/scope/mod.rs:1225
    in find_local_type with name: A#2
    in find_type with name: A#2
    in register_type
    in Stmt with line_col: "(33:1-33:24)"

  DEBUG Scope.register_type(A#2): false
    at crates/stc_ts_file_analyzer/src/analyzer/scope/mod.rs:624
    in Scope::register_type
    in register_type
    in Stmt with line_col: "(33:1-33:24)"

```

**Comment by kdy1** [comment]
Ah, it's a problem with the log message.

https://github.com/dudykr/stc/blob/1b605ab81770ea9db2d76bf3857d06a5a5d540fc/crates/stc_ts_file_analyzer/src/analyzer/scope/mod.rs#L1164

The function name is `find_local_type`, which is only responsible for finding type in the local scope.

**Comment by awareness481** [comment]
> Ah, it's a problem with the log message.
> 
> https://github.com/dudykr/stc/blob/1b605ab81770ea9db2d76bf3857d06a5a5d540fc/crates/stc_ts_file_analyzer/src/analyzer/scope/mod.rs#L1164
> 
> The function name is `find_local_type`, which is only responsible for finding type in the local scope.

So is it correct to say that the issue is not that the intrinsic type is not found, but given the the following statement

```ts
type X = SomeType
```
in some cases (maybe all), `SomeType` is never checked if it's a valid type.

If that's correct I can update the issue title & description

**Comment by kdy1** [comment]
Yes, I think that's the problem.

**Comment by awareness481** [comment]
I think I've found a solution so you can assign me to this task if you want.

**Comment by kdy1** [comment]
Of course, I did it. Thank you!

**Comment by awareness481** [comment]
@kdy1 before I proceed more I wanted to ask you if there's a way to convert an `RTsType::TsTypeRef` to a `Type`. Currently I'm doing this

```rs
                        if let RTsType::TsTypeRef(type_ref) = t {
                            if let RTsEntityName::Ident(t) = &type_ref.type_name {
                                match &*t.sym {
                                    "Uppercase" | "Lowercase" | "Capitalize" | "Uncapitalize" => {
                                        let tp = *type_ref.clone().type_params.unwrap();
                                        let a: TypeParamInstantiation = tp.into();

                                        return Ok(Type::StringMapping(StringMapping {
                                            span: d.span,
                                            kind: IntrinsicKind::from(&*t.sym),
                                            type_args: a,
                                            metadata: Default::default(),
                                        }));
                                    }
                                    _ => {}
                                }
                            }
                        }
```

but this requires a lot more code to be implemented correctly ( I'm doing `impl From<Box<RTsType>> for Type` for example ), so just want to make sure whether this is the right path or not before I end up writing a lot more code.

**Comment by kdy1** [comment]
You can do `t.validate_with(self)` where `t` is `RTsType` and `self` is `Analyzer`.

**Comment by awareness481** [comment]
Good news is that I think I'm on a good path, the "bad news" is that there are a lot of regressions (or false positives because there were no type-checking). For example this file https://github.com/dudykr/stc/blob/main/crates/stc_ts_type_checker/tests/conformance/types/typeAliases/intrinsicTypes.ts causes 20+ extra errors because the current logic doesn't handle them correctly.

I'll open a draft PR as soon as I'm confident about the logic but will probably take a while. Excited to work on this because it's pretty fun :P

**Comment by kdy1** [comment]
Such regression is fine and glad to hear that you feel it's fun 😄
